File: CinePlayer/CinePlayer/Services/FilenameParser.swift
import Foundation

// MARK: - FilenameParser
/// Parses video filenames to extract movie/series title, year, season, and episode information.
/// Handles common patterns used in media naming conventions (Scene naming, Plex naming, etc.)
struct FilenameParser {

    struct ParsedMediaInfo {
        var title: String
        var year: Int?
        var mediaType: MediaType
        var seasonNumber: Int?
        var episodeNumber: Int?
        var episodeTitle: String?
        var quality: String?      // e.g. "1080p", "4K", "720p"
        var releaseGroup: String? // e.g. "YIFY", "BluRay"
    }

    // MARK: - Public API

    static func parse(filename: String) -> ParsedMediaInfo {
        // Strip the file extension
        let nameWithoutExt = stripExtension(filename)

        // Attempt TV episode patterns first (more specific)
        if let tvInfo = parseTVEpisode(nameWithoutExt) {
            return tvInfo
        }

        // Then try movie patterns
        return parseMovie(nameWithoutExt)
    }

    // MARK: - TV Episode Parsing

    private static func parseTVEpisode(_ name: String) -> ParsedMediaInfo? {
        let normalized = name.replacingOccurrences(of: ".", with: " ")
                              .replacingOccurrences(of: "_", with: " ")

        // Pattern list: (regex, titleGroup, seasonGroup, episodeGroup)
        let patterns: [(String, Int, Int, Int)] = [
            // S01E01 or S01E01E02 (standard)
            (#"^(.*?)[.\s_](?:[Ss](\d{1,2})[Ee](\d{1,2}))"#, 1, 2, 3),
            // 1x01
            (#"^(.*?)[.\s_](\d{1,2})x(\d{2})"#, 1, 2, 3),
            // Season 1 Episode 1
            (#"^(.*?)[.\s_][Ss]eason[.\s_](\d{1,2})[.\s_][Ee]pisode[.\s_](\d{1,2})"#, 1, 2, 3),
            // 101 (no separator, e.g. Show.101)
            (#"^(.*?)[.\s_](\d)(\d{2})[.\s_]"#, 1, 2, 3),
        ]

        for (pattern, titleGroup, seasonGroup, episodeGroup) in patterns {
            if let match = normalized.range(of: pattern, options: .regularExpression) {
                let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive)
                let range = NSRange(normalized.startIndex..., in: normalized)
                if let result = regex?.firstMatch(in: normalized, range: range) {
                    let titleRange = result.range(at: titleGroup)
                    let seasonRange = result.range(at: seasonGroup)
                    let episodeRange = result.range(at: episodeGroup)

                    guard let titleR = Range(titleRange, in: normalized),
                          let seasonR = Range(seasonRange, in: normalized),
                          let episodeR = Range(episodeRange, in: normalized) else { continue }

                    let rawTitle = String(normalized[titleR])
                    let title = cleanTitle(rawTitle)
                    let season = Int(normalized[seasonR]) ?? 1
                    let episode = Int(normalized[episodeR]) ?? 1
                    let quality = extractQuality(normalized)

                    if title.isEmpty { continue }

                    return ParsedMediaInfo(
                        title: title,
                        year: extractYear(normalized),
                        mediaType: .episode,
                        seasonNumber: season,
                        episodeNumber: episode,
                        episodeTitle: nil,
                        quality: quality,
                        releaseGroup: extractReleaseGroup(normalized)
                    )
                }
                // Avoid unused variable warning
                _ = match
            }
        }

        return nil
    }

    // MARK: - Movie Parsing

    private static func parseMovie(_ name: String) -> ParsedMediaInfo {
        let normalized = name.replacingOccurrences(of: ".", with: " ")
                              .replacingOccurrences(of: "_", with: " ")

        let year = extractYear(normalized)
        let quality = extractQuality(normalized)
        let releaseGroup = extractReleaseGroup(normalized)

        // Extract title: everything before the year or quality marker
        var title = normalized
        let qualityKeywords = ["1080p", "720p", "480p", "2160p", "4K", "UHD",
                               "BluRay", "BRRip", "WEB-DL", "WEBRip", "HDRip",
                               "DVDRip", "HDTV", "REMUX", "x264", "x265",
                               "HEVC", "H264", "H265", "AAC", "AC3", "DTS",
                               "PROPER", "REPACK", "EXTENDED", "THEATRICAL"]

        if let y = year {
            let yearPattern = #"\b"# + String(y) + #"\b"#
            if let yearRange = title.range(of: yearPattern, options: .regularExpression) {
                title = String(title[..<yearRange.lowerBound])
            }
        } else {
            // Try to find quality marker position
            for keyword in qualityKeywords {
                if let range = title.range(of: keyword, options: .caseInsensitive) {
                    let candidate = String(title[..<range.lowerBound])
                    if !candidate.trimmingCharacters(in: .whitespaces).isEmpty {
                        title = candidate
                        break
                    }
                }
            }
        }

        return ParsedMediaInfo(
            title: cleanTitle(title),
            year: year,
            mediaType: .movie,
            seasonNumber: nil,
            episodeNumber: nil,
            episodeTitle: nil,
            quality: quality,
            releaseGroup: releaseGroup
        )
    }

    // MARK: - Helpers

    static func extractYear(_ text: String) -> Int? {
        // Match 4-digit years between 1900 and 2099
        let pattern = #"\b(19[0-9]{2}|20[0-9]{2})\b"#
        let regex = try? NSRegularExpression(pattern: pattern)
        let range = NSRange(text.startIndex..., in: text)
        if let match = regex?.firstMatch(in: text, range: range),
           let r = Range(match.range(at: 1), in: text) {
            return Int(text[r])
        }
        return nil
    }

    static func extractQuality(_ text: String) -> String? {
        let qualities = ["2160p", "4K", "UHD", "1080p", "720p", "480p", "360p"]
        let lowered = text.lowercased()
        for q in qualities {
            if lowered.contains(q.lowercased()) {
                return q
            }
        }
        return nil
    }

    private static func extractReleaseGroup(_ text: String) -> String? {
        // Release group is usually after the last '-' at the end
        let components = text.components(separatedBy: "-")
        guard let last = components.last else { return nil }
        let cleaned = last.trimmingCharacters(in: .whitespaces)
        // Must be short and alphanumeric
        if cleaned.count > 1 && cleaned.count <= 15 &&
           cleaned.range(of: #"^[A-Za-z0-9]+"#, options: .regularExpression) != nil {
            return cleaned
        }
        return nil
    }

    static func cleanTitle(_ raw: String) -> String {
        var title = raw.trimmingCharacters(in: .whitespaces)
        // Remove trailing/leading punctuation and spaces
        title = title.trimmingCharacters(in: CharacterSet(charactersIn: ".-_() "))
        // Replace multiple spaces with single
        while title.contains("  ") {
            title = title.replacingOccurrences(of: "  ", with: " ")
        }
        // Capitalize words
        return title.split(separator: " ")
            .map { word -> String in
                let w = String(word)
                // Keep articles lowercase unless first word
                let articles = ["a", "an", "the", "of", "in", "on", "at", "to", "for", "and", "or", "but"]
                if articles.contains(w.lowercased()) && title.split(separator: " ").first.map(String.init) != w {
                    return w.lowercased()
                }
                return w.prefix(1).uppercased() + w.dropFirst().lowercased()
            }
            .joined(separator: " ")
    }

    private static func stripExtension(_ filename: String) -> String {
        let videoExtensions = ["mp4", "mkv", "avi", "mov", "m4v", "wmv", "flv", "webm",
                               "ts", "m2ts", "mpg", "mpeg", "3gp", "ogv"]
        let url = URL(fileURLWithPath: filename)
        if videoExtensions.contains(url.pathExtension.lowercased()) {
            return url.deletingPathExtension().lastPathComponent
        }
        return url.lastPathComponent
    }
}

File: CinePlayerAndroid/app/src/main/java/com/cineplayer/android/services/FilenameParser.kt
package com.cineplayer.android.services

import com.cineplayer.android.models.MediaType

data class ParsedMediaInfo(
    val title: String,
    val year: Int? = null,
    val mediaType: MediaType = MediaType.UNKNOWN,
    val seasonNumber: Int? = null,
    val episodeNumber: Int? = null,
    val episodeTitle: String? = null,
    val quality: String? = null
)

object FilenameParser {
    private val QUALITY_MARKERS = listOf(
        "2160p", "4K", "UHD", "1080p", "1080i", "720p", "480p", "360p",
        "BluRay", "BRRip", "BDRip", "WEB-DL", "WEBRip", "HDTV", "DVDRip",
        "x264", "x265", "HEVC", "H264", "H265", "AAC", "AC3", "DTS", "HDR"
    )

    private val SE_PATTERN = Regex("""[Ss](\d{1,2})[Ee](\d{1,2})""")
    private val X_PATTERN = Regex("""(\d{1,2})x(\d{2})""", RegexOption.IGNORE_CASE)
    private val SEASON_EP_PATTERN = Regex("""[Ss]eason\s*(\d+)\s*[Ee]pisode\s*(\d+)""", RegexOption.IGNORE_CASE)
    private val YEAR_PATTERN = Regex("""(?<!\d)(19|20)\d{2}(?!\d)""")

    fun parse(filename: String): ParsedMediaInfo {
        val name = filename.substringBeforeLast(".")
        val quality = extractQuality(name)

        SE_PATTERN.find(name)?.let { match ->
            val season = match.groupValues[1].toInt()
            val episode = match.groupValues[2].toInt()
            val title = cleanTitle(name.substring(0, match.range.first))
            return ParsedMediaInfo(title = title, mediaType = MediaType.EPISODE,
                seasonNumber = season, episodeNumber = episode, quality = quality)
        }

        X_PATTERN.find(name)?.let { match ->
            val season = match.groupValues[1].toInt()
            val episode = match.groupValues[2].toInt()
            val title = cleanTitle(name.substring(0, match.range.first))
            return ParsedMediaInfo(title = title, mediaType = MediaType.EPISODE,
                seasonNumber = season, episodeNumber = episode, quality = quality)
        }

        SEASON_EP_PATTERN.find(name)?.let { match ->
            val season = match.groupValues[1].toInt()
            val episode = match.groupValues[2].toInt()
            val title = cleanTitle(name.substring(0, match.range.first))
            return ParsedMediaInfo(title = title, mediaType = MediaType.EPISODE,
                seasonNumber = season, episodeNumber = episode, quality = quality)
        }

        val year = extractYear(name)
        val title = if (year != null) cleanTitle(name.substringBefore(year.toString()))
                    else cleanTitle(name)
        return ParsedMediaInfo(title = title.ifBlank { name }, year = year,
            mediaType = MediaType.MOVIE, quality = quality)
    }

    private fun extractYear(text: String): Int? = YEAR_PATTERN.find(text)?.value?.toIntOrNull()

    private fun extractQuality(text: String): String? =
        QUALITY_MARKERS.firstOrNull { text.contains(it, ignoreCase = true) }

    fun cleanTitle(raw: String): String {
        var cleaned = raw
        for (marker in QUALITY_MARKERS) {
            cleaned = cleaned.replace(Regex("""\b${Regex.escape(marker)}\b""", RegexOption.IGNORE_CASE), " ")
        }
        cleaned = cleaned.replace(Regex("""[._\-]+"""), " ")
        cleaned = cleaned.replace(Regex("""\[.*?]|\(.*?\)"""), " ")
        cleaned = cleaned.trim().replace(Regex("""\s+"""), " ")
        return toTitleCase(cleaned)
    }

    private fun toTitleCase(text: String): String {
        val articles = setOf("a", "an", "the", "and", "but", "or", "in", "on", "at", "to", "for", "of", "with")
        return text.split(" ").mapIndexed { index, word ->
            if (index == 0 || word.lowercase() !in articles)
                word.replaceFirstChar { it.uppercaseChar() }
            else word.lowercase()
        }.joinToString(" ")
    }

    fun normalizeName(name: String): String =
        name.lowercase().replace(Regex("""^the\s+"""), "").replace(Regex("""[^a-z0-9]"""), "")
}
